core/src/main/java/hudson/Util.java

```diff
@@ -781,15 +785,15 @@ public static String ensureEndsWith(@CheckForNull String subject, @CheckForNull
     public static String getDigestOf(@Nonnull InputStream source) throws IOException {
         try {
             MessageDigest md5 = MessageDigest.getInstance("MD5");
-
-            byte[] buffer = new byte[1024];
-            try (DigestInputStream in = new DigestInputStream(source, md5)) {
-                while (in.read(buffer) >= 0)
-                    ; // simply discard the input
-            }
+            DigestInputStream in = new DigestInputStream(source, md5);
+            // Note: IOUtils.copy() buffers the input internally, so there is no
+            // need to use a BufferedInputStream.
+            IOUtils.copy(in, NullOutputStream.NULL_OUTPUT_STREAM);
             return toHexString(md5.digest());
         } catch (NoSuchAlgorithmException e) {
             throw new IOException("MD5 not installed",e);    // impossible
+        } finally {
+            source.close();
```

Comment:
It is a bugfix, the behavior actually violates Javadoc. Could you please submit a JIRA ticket?
It seems @jglick removed that in the fix for https://issues.jenkins-ci.org/browse/JENKINS-18178 (see the commented code below). https://github.com/jenkinsci/jenkins/commit/e0a3a1ddf5fa08e80a94a565dd05c39f76de6e93

Reply:
The previous code used try-with-resources `try (DigestInputStream in = new DigestInputStream(source, md5))`, which guarantees the underlying stream is closed, right? [e0a3a1d](https://github.com/jenkinsci/jenkins/blob/e0a3a1ddf5fa08e80a94a565dd05c39f76de6e93/core/src/main/java/hudson/Util.java#L568) has a finally block which closes `in`, which itself will close `source`, and the code was later updated in [7b39424](https://github.com/jenkinsci/jenkins/blob/7b3942449516b564ede642b85b99ccc83063c8ff/core/src/main/java/hudson/Util.java#L768) to use try-with-resources to close `in`.

My only concern is if there is an issue here with only closing `source` and not `in`.

Reply:
Yes, the closing of `source` was usually accomplished by closing the `DigestInputStream in = new DigestInputStream(source, md5)` resource (implicitly, as part of the try-with-resource's automatic finally block which would close the resource, which in turn would close the wrapped `InputStream` `source`).

I don't think that there is a problem with not closing the `DigestInputStream`. Examining the OpenJDK sources, `DigestInputStream` does not override close(). `DigestInputStream` extends `FilterInputStream`, whose close() implementation only closes the wrapped `InputStream`:

```java
    public void close() throws IOException {
        in.close();
    }
```

Reply:
right. Sorry, I should have read the code more carefully